tests/test_tcp_server.cpp

```diff
@@ -0,0 +1,341 @@
+        message_callback_ == true)
+    {
+      message_callback_ = false;
+      return true;
+    }
+    else
+    {
+      return false;
+    }
+  }
+
+  int port_ = 50001;
+  std::string message_ = "";
+  int client_fd_ = -1;
+
+private:
+  std::condition_variable connect_cv_;
+  std::mutex connect_mutex_;
+
+  std::condition_variable disconnect_cv_;
+  std::mutex disconnect_mutex_;
+
+  std::condition_variable message_cv_;
+  std::mutex message_mutex_;
+
+  bool connection_callback_ = false;
+  bool disconnection_callback_ = false;
+  bool message_callback_ = false;
+};
+
+TEST_F(TCPServerTest, socket_creation)
+{
+  comm::TCPServer server(port_);
+
+  // Shouldn't be able to create antoher server on same port
+  EXPECT_THROW(comm::TCPServer server2(port_), std::system_error);
+
+  // Shouldn't be able to start the worker thread, when no callbacks has been registered
+  EXPECT_FALSE(server.start());
+
+  server.setMessageCallback(std::bind(&TCPServerTest_socket_creation_Test::messageCallback, this, std::placeholders::_1,
+                                      std::placeholders::_2));
+  server.setConnectCallback(
+      std::bind(&TCPServerTest_socket_creation_Test::connectionCallback, this, std::placeholders::_1));
+  server.setDisconnectCallback(
+      std::bind(&TCPServerTest_socket_creation_Test::disconnectionCallback, this, std::placeholders::_1));
+
+  EXPECT_TRUE(server.start());
+}
+
+TEST_F(TCPServerTest, callback_functions)
+{
+  comm::TCPServer server(port_);
+  server.setMessageCallback(std::bind(&TCPServerTest_callback_functions_Test::messageCallback, this,
+                                      std::placeholders::_1, std::placeholders::_2));
+  server.setConnectCallback(
+      std::bind(&TCPServerTest_callback_functions_Test::connectionCallback, this, std::placeholders::_1));
+  server.setDisconnectCallback(
+      std::bind(&TCPServerTest_callback_functions_Test::disconnectionCallback, this, std::placeholders::_1));
+  server.start();
+
+  // Check that the appropriate callback functions are called
+  Client client(port_);
+  EXPECT_TRUE(waitForConnectionCallback());
+
+  client.send("message\n");
+  EXPECT_TRUE(waitForMessageCallback());
+
+  client.close();
+  EXPECT_TRUE(waitForDisconnectionCallback());
+}
+
+TEST_F(TCPServerTest, unlimited_clients_allowed)
+{
+  comm::TCPServer server(port_);
+  server.setMessageCallback(std::bind(&TCPServerTest_unlimited_clients_allowed_Test::messageCallback, this,
+                                      std::placeholders::_1, std::placeholders::_2));
+  server.setConnectCallback(
+      std::bind(&TCPServerTest_unlimited_clients_allowed_Test::connectionCallback, this, std::placeholders::_1));
+  server.setDisconnectCallback(
+      std::bind(&TCPServerTest_unlimited_clients_allowed_Test::disconnectionCallback, this, std::placeholders::_1));
+  server.start();
+
+  // Test that unlimited number of clients can connect
```

Comment:
3 == unlimited? :laughing:

Reply:
I will change the comment and the test name :laughing: